### PR TITLE
mpl: Fix an IFS-related configure script error

### PR DESCRIPTION
**Pull request by chissg**
## Pull Request Description

In the case where someone specifies `--enable-g=debug` (say) a warning is triggered in `mpl/configure`. However, since `IFS=","` at that point, two cryptic shell errors are produced instead. This PR temporarily pops `IFS` to allow the warnings to succeed. Ideally one would also filter-out `enable_g` flags understood at the higher level but not here before `mpl/configure` is invoked, but I believe that is a policy decision beyond the purview of a first-time contributor.

## Expected Impact

Minor: none beyond fixing the described problem.

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Remove xfail from the test suite when fixing a test
* [ ] Commits are self-contained and do not do two things at once
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Passes whitespace checkers
* [ ] Passes warning tests
* [ ] Passes all tests
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] You or your company has a signed contributor's agreement on file with Argonne
* [ ] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by hzhou** [comment]
> at that point, two cryptic shell errors are produced instead. 

For reference and save some bother, could you paste an example error messages?

**Comment by chissg** [comment]
@hzhou Thanks for the response. As requested:

From the configure command:
```console
$ env CC=gcc CXX=g++ FC=gfortran MPICH_CXXFLAGS=-Wno-deprecated-declarations 'CFLAGS=-D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64' stdver=c++17 /scratch/greenc/test-products/mpich/v3_3_2a/source/mpich-3.3.2/configure --prefix=/scratch/greenc/test-products/mpich/v3_3_2a/Linux64bit+3.10-2.17-e20-prof --enable-fortran=all --enable-cxx --enable-romio --enable-debuginfo --enable-versioning --enable-threads=runtime --disable-wrapper-rpath --enable-shared --disable-static --disable-libxml2 --enable-timer-type=clock_gettime --enable-g=debug --disable-dependency-tracking --disable-maintainer-mode --enable-mpit-pvars=none --enable-strict --enable-timing=none --enable-fast=ndebug,O3 > configure.log 2>&1
```
we get:
```console
<snip>
checking if C compiler supports __builtin_expect... yes
/scratch/greenc/test-products/mpich/v3_3_2a/source/mpich-3.3.2/src/mpl/configure: line 12600: printf %s\n: command not found
/scratch/greenc/test-products/mpich/v3_3_2a/source/mpich-3.3.2/src/mpl/configure: line 12601: printf %s\n: command not found
checking for gcov... gcov
checking whether the compiler supports __typeof(variable)... yes
<snip>
```
Expected output:
```console
<snip>
checking if C compiler supports __builtin_expect... yes
configure: WARNING: Unknown value debug for enable-g
checking for gcov... gcov
checking whether the compiler supports __typeof(variable)... yes
<snip>
```
with a corresponding message (previously absent) in `src/mpl/config.log`:
```console
configure:12600: WARNING: Unknown value debug for enable-g
```

**Comment by chissg** [comment]
FYI, and to whom it may concern: CLA signed and sent by FNAL legal office to appropriate email list. Please let me know if anything else needs to be done on the bureaucratic side to enable this PR to be accepted.

**Comment by hzhou** [comment]
@chissg Thanks for the explanation, I understand now.

There is similar fix in the main `configure.ac`: https://github.com/pmodels/mpich/blob/3db9b52bf24d6b671b78a897df93ff1d0c5df1a2/configure.ac#L1537-L1540

Could you do the exact fix for consistency? To help the code clearer, could you add a comment (to both places): `# default IFS is needed for AC_MSG_WARN` ?

The root issue is `IFS` has side effects that typically we are not aware, so I think the right fix is to parse the comma separated list without altering IFS. We'll do that separately.

**Comment by chissg** [comment]
The nature of the `IFS` handling in the two `configure.ac` files is different: in the example you cite, the `IFS` handling is manual in every case, and therefore the accommodation you show is consistent with the rest of the file. In `src/mpl/configure.ac`, `IFS` is changed elsewhere in the file with the aid of the `PAC_PUSH_FLAG()` and `PAC_POP_FLAG()` macros that explicitly handle nesting, so my patch was intended to be consistent with existing use _within that file_—see https://github.com/pmodels/mpich/blob/3db9b52bf24d6b671b78a897df93ff1d0c5df1a2/src/mpl/configure.ac#L82-L83 and https://github.com/pmodels/mpich/blob/3db9b52bf24d6b671b78a897df93ff1d0c5df1a2/src/mpl/configure.ac#L107

**TL;DR**: I can certainly make the change you request, but are you sure you want me to?

**Comment by hzhou** [comment]
> **TL;DR**: I can certainly make the change you request, but are you sure you want me to?

I guess either is  fine with me, so I'll defer to your decision. Could you  add the comment I mentioned?

**Comment by chissg** [comment]
> > **TL;DR**: I can certainly make the change you request, but are you sure you want me to?
> 
> I guess either is fine with me, so I'll defer to your decision. Could you add the comment I mentioned?

No problem: incoming shortly.

**Comment by chissg** [comment]
@hzhou Done. Given that you'd mentioned changing the way `enable_g` is processed in the future, I now save the `IFS` value required to process `enable_g` at the beginning of the section (with an appropriate comment) to provide a single point of maintenance for the specified IFS value and cementing the connection between the initial setting and the accommodations needed around `AC_MSG_WARN`. Let me know if this works for you, or if I should revise.

**Comment by hzhou** [comment]
> FYI, and to whom it may concern: CLA signed and sent by FNAL legal office to appropriate email list. Please let me know if anything else needs to be done on the bureaucratic side to enable this PR to be accepted.

@raffenet Could you check the CLA situation?

**Comment by raffenet** [comment]
Verified the FNAL CLA. I have created a team and invited @chissg. Will rebase/merge once he accepts.

**Comment by hzhou** [comment]
> 
> 
> Verified the FNAL CLA. I have created a team and invited @chissg. Will rebase/merge once he accepts.

Sounds good 👍 For occasional contributors, the github team feature may not be that useful. An alternative is to simply add the approved contributors to our `jenkins-scripts/mpich_allowlist.txt` -- once the CLA is verified.
